### ISSUE TITLE
Calling sign_unplace() from CursorMoved causes line below Visual area to be redrawn incorrectly when moving cursor above first screen line with scrolloff=0

### ISSUE THREAD

**Issue by zeertzjq** [vim/vim#10017]
### Steps to reproduce

1. Run `vim --clean`
2. Source the following file:
```
set lines=8 scrolloff=0

call setline(1, repeat(['foo'], 10))
call sign_define('foo', { 'text': '>' })
call sign_place(1, 'bar', 'foo', bufnr(), { 'lnum': 2 })
call sign_place(2, 'bar', 'foo', bufnr(), { 'lnum': 1 })

autocmd CursorMoved * if getcurpos()[1] == 2 | call sign_unplace('bar', { 'id': 1 }) | endif
```
3. Press `VG7kk`
4. A part of the second screen still has Visual highlight

### Expected behaviour

The second screen does not have Visual highlight

### Version of Vim

8.2.4622

### Environment

Operating system: Arch Linux
Terminal: kitty
Value of $TERM: xterm-kitty
Shell: fish

**Comment by zeertzjq** [comment]
Reverting the code change from patch 8.2.1488 fixes the bug:
```diff
diff --git a/src/drawscreen.c b/src/drawscreen.c
index a902397cd..12eed8c47 100644
--- a/src/drawscreen.c
+++ b/src/drawscreen.c
@@ -1730,7 +1730,7 @@ win_update(win_T *wp)
 	if (mod_top != 0
 		&& wp->w_topline == mod_top
 		&& (!wp->w_lines[0].wl_valid
-		    || wp->w_topline <= wp->w_lines[0].wl_lnum))
+		    || wp->w_topline == wp->w_lines[0].wl_lnum))
 	{
 	    // w_topline is the first changed line and window is not scrolled,
 	    // the scrolling from changed lines will be done further down.
```
Reverting this change does not seem to cause the test added in that patch to fail. That patch also indeed looks very strange, as the comment below says `window is not scrolled`, but when `wp->w_topline < wp->w_lines[0].wl_lnum` the window is scrolled.

**Comment by zeertzjq** [comment]
Oh, it look [patch 8.2.3204](https://github.com/vim/vim/commit/c9e7e344ed390d2a22afb88001b6aa80832d2541) also fixes #6739 with patch 8.2.1488 reverted.

**Comment by zeertzjq** [comment]
I guess the bug described in #6739 happens because the lines scrolled back up again after being scrolled down. patch 8.2.1488 fixes that by preventing them from being scrolled down, while patch 8.2.3204 fixes that by prevent them from being scrolled back up again. As the window has indeed been scrolled when `wp->w_topline < wp->wp->w_lines[0].wl_lnum`, I guess patch 8.2.1488 was wrong, and patch 8.2.3204 was the correct fix for the bug described in #6739.
